File: 023/023.py
# coding: utf-8
import itertools
import math

def get_sum_of_divisors(n):
    sqrt = math.sqrt(n)
    _sum = 0
    for i in range(1, math.trunc(sqrt)+1):
        if i == 1:
            _sum += 1
        elif n%i == 0:
            if i**2 == n:
                _sum += i
            else:
                _sum += i + n/i
    return _sum

ans = 0
for i in range(2, 28123):
    a = get_sum_of_divisors(i)
    if i < a:
        ans += i

print(ans)

File: 023/023.exs
sum = fn(n) ->
        sqrt = :math.sqrt(n)
        Enum.reduce 1..(trunc(sqrt)), fn(x, sum) ->
          cond do
            x == 1 ->
              sum + x
            rem(n, x) == 0 && :math.pow(x, 2) == n ->
              sum + x
            rem(n, x) == 0 ->
              sum + x + div(n, x)
            true ->
              sum
          end
        end
      end

1..28123 |> Enum.filter(fn(x) -> sum.(x) < x end) |> Enum.reduce(0, &+/2) |> IO.puts
